### ISSUE TITLE
How do I implement a SSL Certificate Chain File for Bitnami's Apache2 running in a docker container

### ISSUE THREAD

**Issue by nolssmit** [bitnami/bitnami-docker-apache#105]
Hi,

I'm using a subdomain from [No-IP](https://my.noip.com) and bought their proposed SSL certificate from [TrustCor](https://trustcor.com) and installed it by:

- Mapping the certs folder of Apache running in a docker container (see my docker-compose.yml file below) to my persistent apache-certs folder
- Following the steps in Binami's Apache documentation:  [Using custom SSL certificates](https://github.com/bitnami/bitnami-docker-apache)
- Renamed my downloaded PEM Chain file (slimzulu.hopto.org.pem-chain) to server-ca.crt, PEM file (slimzulu.hopto.org.pem) to server.crt and my slimzulu.hopto.org.key to server.key and copied them to my persistent apache-certs folder

I then followed the steps explained by [TrustCor](https://trustcor.com/support/ssl-install/apache-ssl-installation.php) to modify the virtual host file to implement the SSL Certificate Chain File (see my my-vhost.conf file below).

but if I use the service from [SSLShopper](https://www.sslshopper.com/ssl-checker.html#hostname=slimzulu.hopto.org) to verify the implementation, I get the following response:

> The certificate is not trusted in all web browsers. You may need to install an Intermediate/chain certificate to link it to a trusted root certificate. Learn more about this error. The fastest way to fix this problem is to contact your SSL provider.

**The relevant part of my docker-compose.yml file**
```
  apache:
    restart: unless-stopped
    container_name: apache
    image: bitnami/apache:latest
    environment:
      - APACHE_HTTP_PORT_NUMBER=8080 
      - APACHE_HTTPS_PORT_NUMBER=8443
    ports:
      - 80:8080
      - 443:8443
    volumes:
      - ./apache_data:/app
      - ./apache-certs:/certs
      - .my_vhost.conf:/vhosts/my_vhost.conf:ro
```

**My my_vhost.conf file**
```
<VirtualHost *:8080>
  ServerName slimzulu.hopto.org
  SSLCertificateChainFile ./apache-certs/server-ca.crt

  DocumentRoot "/app"
  <Directory "/app">
    Options Indexes FollowSymLinks
    AllowOverride All
    Require all granted
  </Directory>
</VirtualHost>
```

Bitnami's Apache documentation is not clear how to implement a SSL Certificate Chain File for Apache running in a docker container.

Help will be highly appreciated to resolve this issue.

Regards,

Nols Smit

**Comment by javsalgar** [comment]
Hi,

Could you confirm that, when you access the application with your browser, you get the certificate you mounted or a different one?

**Comment by nolssmit** [comment]
Hi Javier,

The first time I accessed the web site (https://slimzulu.hopto.org), I had
to confirm to my antivirus program (Bitdefender) that I trust the web
site.  You can test it yourself.

After opening the web page and checking the SSL certificate, using Chrome,
I do see my SSL certificate is being used but the root certificate is that
of Bitdefender and not that of TrustCor, the issuer of my SSL certificate.
[image: image.png]

My my_vhost.conf file:
[image: image.png]

The response on:  $ docker ps -a
[image: image.png]

Regards,

Nols Smit

**Comment by javsalgar** [comment]
Hi,

I tried entering and what I see is this certificate

![image](https://user-images.githubusercontent.com/1133777/123755897-acd4a380-d8bc-11eb-92a0-42608c4b49aa.png)


Is this the one you installed? Just to confirm if this is expected.

**Comment by nolssmit** [comment]
Hi Javier,

No, that's not the one.

The correct certificate is for slimzulu.org.za, issued by TrustCor, and the
certificate info should show:

-----BEGIN CERTIFICATE-----
MIIGnjCCBYagAwIBAgIMBgFFo7PgxedI8J8aMA0GCSqGSIb3DQEBCwUAMFsxCzAJ
BgNVBAYTAlBBMSQwIgYDVQQKDBtUcnVzdENvciBTeXN0ZW1zIFMuIGRlIFIuTC4x
JjAkBgNVBAMMHVRydXN0Q29yIERWIFNTTCBDQSAtIEcyIC0gUlNBMB4XDTIxMDYy
MzE2MjIxNloXDTIyMDQyMTE2MjIxNVowHTEbMBkGA1UEAwwSc2xpbXp1bHUuaG9w
dG8ub3JnMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAscNMXfpxcOST
Ws6OD6qq8ThK2nCxwQrn0gBsFiN1/22L4fa83f2+8SSLa6z2KHnaMf+5BtEmr/ga
kXuYlLuo9TfDNqqH2TpoxHToAprXilR9hCe+u2OvIRjJVpHQmSMllCCWmcPHsJd1
PDN80mtKzRCqNyTgFYAA7Pc+WOZcg/urXj8kAD80KDAZDWGcfLdAcyFp1wAwrfpS
BZYapET9EeZ7zAag6NNTI+n1bspOb+tbyCa8lCohj1yh5CsETNmAwNZeIU9NzW8i
3faeCxiPqerH2gUmXOqYqQfREjH1JDrkTaGqMrvCYF262hG4sgJiq5cTLZs/Z+CZ
BEi770/ZaQIDAQABo4IDnjCCA5owbAYIKwYBBQUHAQEEYDBeMCMGCCsGAQUFBzAB
hhdodHRwOi8vb2NzcC50cnVzdGNvci5jYTA3BggrBgEFBQcwAoYraHR0cDovL2Nl
cnRzLnRydXN0Y29yLmNhL3N1Yi9kdi1zc2wtcnNhLmRlcjAfBgNVHSMEGDAWgBQS
7B6ctP/DJBeLnvC8cNUeL2B0VzAJBgNVHRMEAjAAMD8GA1UdHwQ4MDYwNKAyoDCG
Lmh0dHA6Ly9jcmwudHJ1c3Rjb3IuY2Evc3ViL2R2LXNzbC1yc2Etcy0yNi5jcmww
HQYDVR0lBBYwFAYIKwYBBQUHAwIGCCsGAQUFBwMBMB0GA1UdDgQWBBQ0Q9RIedie
FtFDnn/hh+tIyy+S0zAOBgNVHQ8BAf8EBAMCBaAwHQYDVR0RBBYwFIISc2xpbXp1
bHUuaG9wdG8ub3JnMFcGA1UdIARQME4wQgYLKwYBBAGC138BAQ0wMzAxBggrBgEF
BQcCARYlaHR0cDovL3RydXN0Y29yLmNvbS9yZXNvdXJjZXMvY3BzLnBkZjAIBgZn
gQwBAgEwggH1BgorBgEEAdZ5AgQCBIIB5QSCAeEB3wB2AEalVet1+pEgMLWiiWn0
830RLEF0vv1JuIWr8vxw/m1HAAABejmwpy4AAAQDAEcwRQIgPLBQ+Fg/mseuu/JZ
QNKJkD9ZUHYT3tO8Q9hrYzysbjMCIQCFyTG4XBnGLu7REuW4RixecyA2hnjVx40H
7jW/x10TmgB1ACl5vvCeOTkh8FZzn2Old+W+V32cYAr4+U1dJlwlXceEAAABejmw
qWMAAAQDAEYwRAIgOWcahEF3oHGZ7sYhi/zeQ/J/1cEV55MzDfb2Bw/ki8QCICQ3
bbDPBOPv9NJ6Xhc4YJbnxZUtMs9Qq2REFrAJ0UkgAHYAb1N2rDHwMRnYmQCkURX/
dxUcEdkCwQApBo2yCJo32RMAAAF6ObCsvgAABAMARzBFAiEA0r9mWxdg+A6tSt57
IXzxUWJEesRZiaAGHuzEh/g5OCsCIFbgiwPi53EHB9KuL6cze4jsk3AKDi7asItr
dJ80rCLnAHYAQcjKsd8iRkoQxqE6CUKHXk4xixsD6+tLx2jwkGKWBvYAAAF6ObCt
gQAABAMARzBFAiEAiAOgyZv58RYx1JMRgaJPvrNeTIvAiENVFbMFDoI2BDwCIEXv
Jfdj7WHZ39eF09jGTNL3eABAAFO1Rh8LRDmWmn4mMA0GCSqGSIb3DQEBCwUAA4IB
AQAXBjGm7LQDD3wtpk2/DNKvXj/zGlyx5FkvpA0ouqdtjtUMsF4gJ5QgDXJA5rlh
qziEmyBGWrrVS/AwxgUKlE745be30k7wqFWkpjTW7+6nRllPkuOnVeoSi1EuXoOw
/59pVr8qiC3yQnKCV8XmLWbY/w3nC4BqAGHlzjpPMAK+F5zO4AzhlpJ/wNd7FFgR
RkMLYHoSmy5Njju8Uva6mKUkcJMbm6mDzOmwb4ZLVR9BxIeIE4aswrPqnyJnmqfX
Rq5AUdkvFF4HUI7u2BrZ/+N4oVJHYgHoLli59RzF8GRJ8W3ixoJR/ZfUpKj1kzU+
aagFhXxh7nSktn9e4DL/vyTL
-----END CERTIFICATE-----

*The latest version of my my_vhost.conf*
<VirtualHost *:8443>
  ServerName slimzulu.hopto.org

  DocumentRoot "/app"
  <Directory "/app">
    Options Indexes FollowSymLinks
    AllowOverride All
    Require all granted
  </Directory>
</VirtualHost>

*The latest version of the Apache section of my docker-compose.yml*
  apache:
    restart: unless-stopped
    container_name: apache
    image: bitnami/apache:latest
    environment:
      - APACHE_HTTP_PORT_NUMBER=8080
      - APACHE_HTTPS_PORT_NUMBER=8443
    ports:
      - 80:8080
      - 443:8443
    volumes:
      - ./apache_data:/app
      - ./apache-certs:/certs
      - .my_vhost.conf:/vhosts/my_vhost.conf:ro
#      - .httpd.conf:/opt/bitnami/apache/bin/hppd/httpd.conf

*My Apache logs*
***@***.***:~/www$ docker logs apache
apache 13:09:30.70
apache 13:09:30.70 Welcome to the Bitnami apache container
apache 13:09:30.71 Subscribe to project updates by watching
https://github.com/bitnami/bitnami-docker-apache
apache 13:09:30.71 Submit issues and feature requests at
https://github.com/bitnami/bitnami-docker-apache/issues
apache 13:09:30.72
apache 13:09:30.72 INFO  ==> ** Starting Apache setup **
apache 13:09:30.80 INFO  ==> Found mounted virtual hosts in '/vhosts'.
Copying them to '/opt/bitnami/apache/conf/vhosts'
apache 13:09:30.83 INFO  ==> Mounting certificates files from '/certs'
apache 13:09:30.87 INFO  ==> Mounting application files from '/app'
apache 13:09:30.89 INFO  ==> Configuring the HTTP port
apache 13:09:30.98 INFO  ==> Configuring the HTTPS port
apache 13:09:31.02 INFO  ==> ** Apache setup finished! **

apache 13:09:31.09 INFO  ==> ** Starting Apache **
[Tue Jun 29 13:09:31.205144 2021] [ssl:warn] [pid 1] AH01909:
www.example.com:8443:0 server certificate does NOT include an ID which
matches the server name
[Tue Jun 29 13:09:31.248347 2021] [ssl:warn] [pid 1] AH01909:
www.example.com:8443:0 server certificate does NOT include an ID which
matches the server name
[Tue Jun 29 13:09:31.259173 2021] [mpm_prefork:notice] [pid 1] AH00163:
Apache/2.4.48 (Unix) OpenSSL/1.1.1d configured -- resuming normal operations
[Tue Jun 29 13:09:31.259555 2021] [core:notice] [pid 1] AH00094: Command
line: '/opt/bitnami/apache/bin/httpd -f /opt/bitnami/apache/conf/httpd.conf
-D FOREGROUND'
***@***.***:~/www$


Can you test the URL again because I made some modifications in
Dreamfactory.

Regards,

Nols Smit

**Comment by javsalgar** [comment]
Hi,

Could you try editing your vhost configuration and add also these

```
  SSLEngine on
  SSLCertificateFile "/certs/server.crt"
  SSLCertificateKeyFile "/certs/server.key"
  SSLCertificateChainFile "/certs/server-ca.crt"

```

**Comment by nolssmit** [comment]
Hi Javier,

Below is a image of my directory outlay:
[image: image.png]

So my question is, in my vhost file, shouldn't I rather use:

  SSLEngine on
  SSLCertificateFile "/apache-certs/server.crt"
  SSLCertificateKeyFile "/apache-certs/server.key"
  SSLCertificateChainFile "/apache-certs/server-ca"

Regards,

Nols Smit

**Comment by nolssmit** [comment]
Hi Javier,

*The Apache section of my docker-compose.yml file*

  apache:
    restart: unless-stopped
    container_name: apache
    image: bitnami/apache:latest
    environment:
      - APACHE_HTTP_PORT_NUMBER=8080
      - APACHE_HTTPS_PORT_NUMBER=8443
    ports:
      - 80:8080
      - 443:8443
    volumes:
      - ./apache_data:/app
#      - ./apache-certs:/certs
      - .my_vhost.conf:/vhosts/my_vhost.conf:ro

*My my_vhost.conf file*

<VirtualHost *:8080>
  ServerName slimzulu.hopto.org

  SSLEngine on
  SSLCertificateFile "/certs/server.crt"
  SSLCertificateKeyFile "/certs/server.key"
  SSLCertificateChainFile "/certs/server-ca.crt"

  DocumentRoot "/app"
  <Directory "/app">
    Options Indexes FollowSymLinks
    AllowOverride All
    Require all granted
  </Directory>
</VirtualHost>



* I tested with:*

  SSLEngine on
  SSLCertificateFile "/certs/server.crt"
  SSLCertificateKeyFile "/certs/server.key"
  SSLCertificateChainFile "/certs/server-ca.crt"

*and also with:*

  SSLEngine on
  SSLCertificateFile /certs/server.crt
  SSLCertificateKeyFile /certs/server.key
  SSLCertificateChainFile /certs/server-ca.crt


*and also with:*

  SSLEngine on
  SSLCertificateFile "/home/nols/www/apache-certs/server.crt"
  SSLCertificateKeyFile "/home/nols/www/apache-certs/server.key"
  SSLCertificateChainFile "/home/nols/www/apache-certs/server-ca.crt"

*and also with: *

  SSLEngine on
  SSLCertificateFile /home/nols/www/apache-certs/server.crt
  SSLCertificateKeyFile /home/nols/www/apache-certs/server.key
  SSLCertificateChainFile /home/nols/www/apache-certs/server-ca.crt

I even changed the port, in my_vhost.conf, from 8080 to 8443 and re-run the
tests.
I then uncommented " - ./apache-certs:/certs " in my docker-compose.yml
file and re-run the tests with the port equal to 8080 and also with the
port equal to 8443.

After each test, I checked the SSL certificate with DigiCert® SSL
Installation Diagnostics Tool <https://www.digicert.com/help/> and every
result was as per attached file.

Regards,

Nols Smit

**Comment by javsalgar** [comment]
Hi, this is strange

```

  SSLEngine on
  SSLCertificateFile "/home/nols/www/apache-certs/server.crt"
  SSLCertificateKeyFile "/home/nols/www/apache-certs/server.key"
  SSLCertificateChainFile "/home/nols/www/apache-certs/server-ca.crt"


```

This line should make the container crash. It seems like Apache is not reading that file. Could you try using a malformed configuration file and see if Apache crashes? If it doesn't then the file is not being read.

**Comment by nolssmit** [comment]
Hi,

With today's tests:

*The Apache section in my docker-compose.yml file:*
  apache:
    restart: unless-stopped
    container_name: apache
    image: bitnami/apache:latest
    environment:
      - APACHE_HTTP_PORT_NUMBER=8080
      - APACHE_HTTPS_PORT_NUMBER=8443
    ports:
      - 80:8080
      - 443:8443
    volumes:
      - ./apache_data:/app
      - ./apache-certs:/certs
      - .my_vhost.conf:/vhosts/my_vhost.conf:ro


*My normal  my_vhost.conf file:*
<VirtualHost *:8080>
  ServerName slimzulu.hopto.org

  DocumentRoot "/app"
  <Directory "/app">
    Options Indexes FollowSymLinks
    AllowOverride All
    Require all granted
  </Directory>
</VirtualHost>

The SSL certificate is used only when I activate the following line in my
docker-compose.yml file:  - ./apache-certs:/certs

*My malformed my_vhost.conf file:*
<VirtualHost *:99999>
  ServerName slimzuluxxxxx.hoptoxxxxx.orgxxxxx

  DocumentRoot "/appxxxx"
  <Directory "/appxxxx">
    Options Indexes FollowSymLinksxxxxx
    AllowOverride Allxxxxx
    Require all grantedxxxx
  </Directory>
</VirtualHost>

The SSL certificate is used with a malformed  my_vhost.conf file and the
following line in my docker-compose.yml file activated:  -
./apache-certs:/certs
Also, checking wit the DigiCert service:
[image: image.png]


The SSL certificate is *not* used with a malformed  my_vhost.conf file and
the following line in my docker-compose.yml file *de-activated*:  -
./apache-certs:/certs
[image: image.png]

I also think Apache is not reading my_host.conf file


Regards,
Nols Smit

**Comment by javsalgar** [comment]
I think I found the cause

```
      - .my_vhost.conf:/vhosts/my_vhost.conf:ro

```

There's a slash missing, it should be

```
      - ./my_vhost.conf:/vhosts/my_vhost.conf:ro
```

Could you try again?

**Comment by nolssmit** [comment]
Hi,

No, that was not the problem because my_vhost.conf is not a folder.
A working solution is:

*My cocker-compose.yml file:*

# https://hub.docker.com/r/bitnami/dreamfactory/
  dreamfactory:
    restart: unless-stopped
    container_name: dreamfactory
    image: bitnami/dreamfactory:latest
    environment:
      - SMTP_HOST=smtp.gmail.com
      - SMTP_PORT=587
      - ***@***.***
      - SMTP_PASSWORD=xxxxxxxxxxxxxxx
      - SMTP_PROTOCOL=tls
      - DREAMFACTORY_DATABASE_TYPE=mariadb
      - DREAMFACTORY_ENABLE_MARIADB_DATABASE=yes
    ports:
      - 80:80
      - 443:443
    volumes:
      - ./dreamfactory_data:/bitnami
    networks:
      - dreamfactory-tier
    depends_on:
      - mariadb
      - mongodb
      - redis


# https://github.com/bitnami/bitnami-docker-apache
  apache:
    restart: unless-stopped
    container_name: apache
    image: bitnami/apache:latest
    environment:
      - APACHE_HTTP_PORT_NUMBER=8080
      - APACHE_HTTPS_PORT_NUMBER=8443
    ports:
      - 8080:8080
      - 8443:8443
    volumes:
      - ./apache_data:/app
      - ./apache-certs:/certs
      - .my_vhost.conf:/vhosts/my_vhost.conf:ro

*My my_vhost.conf file:*

<VirtualHost *:8080>
  ServerName slimzulu.hopto.org

  DocumentRoot "/app"
  <Directory "/app">
    Options Indexes FollowSymLinks
    AllowOverride All
    Require all granted
  </Directory>
</VirtualHost>


You may test it now by following the link:  https://slimzulu.hopto.org:8443
Then click on the lock left to the web link and query the SSL certificate
info.  Using both DigiCert and SSLShopper gives the wrong info.

My only problem is that I would like to switch host ports between
Dreamfactory and Apache. e.g.
For Dreanfactory use:
    ports:
      - 8080:80
      - 8443:443

For Apache use:
      - 80:8080
      - 443:8443

In my_vhost.conf use:
<VirtualHost *:80>


If I try that, then the certificate chain path is broken.  It is not
natural to specify the port in the daily use of your URL

Regards,

Nols Smit

**Comment by nolssmit** [comment]
Hi Javier,

I'm now satisfied with the working solution regarding the SSL certificate
chain file.

The last problem I have is to use Bitnami's Dreamfactory/Apache docker
images in a real world application.  Attached is an abstract of my manual
(for myself) about the configurations.

I'm using TMS Software Web Core
<https://www.tmssoftware.com/site/tmswebcoreintro.asp> to create a web
application.

But if I run the application to access the database via the Rest interface,
created by Dreamfactory, I get the following error response:

Access to XMLHttpRequest at '
https://slimzulu.hopto.org/api/v2/starsbib/_proc/json_publications_lookups(null,authors,keywords,media,publishers)?api_key=c5c143a1608482641961b9d33203826e637316d915987a640901de73ec807217'
from origin 'https://slimzulu.hopto.org:8443' has been blocked by CORS
policy: Response to preflight request doesn't pass access control check: No
'Access-Control-Allow-Origin' header is present on the requested resource.

But if I execute the command:
https://slimzulu.hopto.org/api/v2/starsbib/_proc/json_publications_lookups(null,authors,keywords,media,publishers)?api_key=c5c143a1608482641961b9d33203826e637316d915987a640901de73ec807217
then I do get a positive response.

I did configure Cors in Dreamfactory, see my attached manual.

Regards,

Nols Smit

**Comment by javsalgar** [comment]
Hi,

Could you create a separate ticket in the bitnami-docker-dreamfactory repository? It seems like two different issues

**Comment by nolssmit** [comment]
will do

**Comment by nolssmit** [comment]
Hi,

I get my own.

[image: image.png]

Regards,

Nols Smit

**Comment by syjung6967** [comment]
I solved this problem temporarily by appending SSLCACertificateFile directive in my /opt/bitnami/apache/conf/vhosts/_app_-https-vhost.conf.
```
+  SSLCACertificateFile "/opt/bitnami/apache/conf/bitnami/certs/server-ca.crt"
   SSLCertificateFile "/opt/bitnami/apache/conf/bitnami/certs/server.crt"
   SSLCertificateKeyFile "/opt/bitnami/apache/conf/bitnami/certs/server.key"
```
Adding the directive for *-https-vhost.conf.tpl files in bitnami-templates would help other people who face this problem.

Corresponding file list:
2.4/debian-10/rootfs/opt/bitnami/scripts/apache/bitnami-templates/app-generic-https-vhost.conf.tpl
2.4/debian-10/rootfs/opt/bitnami/scripts/apache/bitnami-templates/app-proxy-https-vhost.conf.tpl
2.4/debian-10/rootfs/opt/bitnami/scripts/apache/bitnami-templates/app-https-vhost.conf.tpl
2.4/debian-10/rootfs/opt/bitnami/scripts/apache/bitnami-templates/app-ruby-passenger-https-vhost.conf.tpl
2.4/debian-10/rootfs/opt/bitnami/scripts/apache/bitnami-templates/bitnami-ssl.conf.tpl
